File: python/src/bleepstore/metadata/memory.py
"""In-memory metadata store for BleepStore.

Useful for testing and ephemeral deployments. Data is lost on restart.
"""

from datetime import datetime, timezone
from typing import Any


def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.000Z")


class MemoryMetadataStore:
    """In-memory metadata store using Python dicts.

    No persistence - all data is lost on restart. Suitable for testing
    or short-lived ephemeral deployments.
    """

    def __init__(self) -> None:
        self._buckets: dict[str, dict[str, Any]] = {}
        self._objects: dict[tuple[str, str], dict[str, Any]] = {}
        self._uploads: dict[str, dict[str, Any]] = {}
        self._parts: dict[str, dict[int, dict[str, Any]]] = {}
        self._credentials: dict[str, dict[str, Any]] = {}

    async def init_db(self) -> None:
        pass

    async def close(self) -> None:
        self._buckets.clear()
        self._objects.clear()
        self._uploads.clear()
        self._parts.clear()
        self._credentials.clear()

    async def create_bucket(
        self,
        bucket: str,
        region: str = "us-east-1",
        owner_id: str = "",
        owner_display: str = "",
        acl: str = "{}",
    ) -> None:
        if bucket in self._buckets:
            raise KeyError(f"Bucket already exists: {bucket}")
        self._buckets[bucket] = {
            "name": bucket,
            "region": region,
            "owner_id": owner_id,
            "owner_display": owner_display,
            "acl": acl,
            "created_at": _now_iso(),
        }

    async def bucket_exists(self, bucket: str) -> bool:
        return bucket in self._buckets

    async def delete_bucket(self, bucket: str) -> None:
        self._buckets.pop(bucket, None)
        to_delete = [k for k in self._objects if k[0] == bucket]
        for k in to_delete:
            del self._objects[k]

    async def get_bucket(self, bucket: str) -> dict[str, Any] | None:
        return self._buckets.get(bucket)

    async def list_buckets(self, owner_id: str = "") -> list[dict[str, Any]]:
        buckets = list(self._buckets.values())
        if owner_id:
            buckets = [b for b in buckets if b["owner_id"] == owner_id]
        return sorted(buckets, key=lambda b: b["name"])

    async def update_bucket_acl(self, bucket: str, acl: str) -> None:
        if bucket in self._buckets:
            self._buckets[bucket]["acl"] = acl

    async def put_object(
        self,
        bucket: str,
        key: str,
        size: int,
        etag: str,
        content_type: str = "application/octet-stream",
        content_encoding: str | None = None,
        content_language: str | None = None,
        content_disposition: str | None = None,
        cache_control: str | None = None,
        expires: str | None = None,
        storage_class: str = "STANDARD",
        acl: str = "{}",
        user_metadata: str = "{}",
    ) -> None:
        self._objects[(bucket, key)] = {
            "bucket": bucket,
            "key": key,
            "size": size,
            "etag": etag,
            "content_type": content_type,
            "content_encoding": content_encoding,
            "content_language": content_language,
            "content_disposition": content_disposition,
            "cache_control": cache_control,
            "expires": expires,
            "storage_class": storage_class,
            "acl": acl,
            "user_metadata": user_metadata,
            "last_modified": _now_iso(),
            "delete_marker": 0,
        }

    async def object_exists(self, bucket: str, key: str) -> bool:
        return (bucket, key) in self._objects

    async def get_object(self, bucket: str, key: str) -> dict[str, Any] | None:
        return self._objects.get((bucket, key))

    async def delete_object(self, bucket: str, key: str) -> None:
        self._objects.pop((bucket, key), None)

    async def delete_objects_meta(self, bucket: str, keys: list[str]) -> list[str]:
        deleted = []
        for key in keys:
            if (bucket, key) in self._objects:
                del self._objects[(bucket, key)]
                deleted.append(key)
        return deleted

    async def update_object_acl(self, bucket: str, key: str, acl: str) -> None:
        obj = self._objects.get((bucket, key))
        if obj:
            obj["acl"] = acl

    async def list_objects(
        self,
        bucket: str,
        prefix: str = "",
        delimiter: str = "",
        max_keys: int = 1000,
        marker: str = "",
        continuation_token: str | None = None,
    ) -> dict[str, Any]:
        if max_keys <= 0:
            return {
                "contents": [],
                "common_prefixes": [],
                "is_truncated": False,
                "next_continuation_token": None,
                "next_marker": None,
                "key_count": 0,
            }

        start_after = continuation_token or marker or ""

        objects = [obj for (b, _), obj in self._objects.items() if b == bucket]
        objects = sorted(objects, key=lambda o: o["key"])

        if prefix:
            objects = [o for o in objects if o["key"].startswith(prefix)]

        if start_after:
            objects = [o for o in objects if o["key"] > start_after]

        contents: list[dict[str, Any]] = []
        common_prefixes: list[str] = []
        seen_prefixes: set[str] = set()

        for obj in objects:
            if len(contents) + len(common_prefixes) >= max_keys:
                break

            key: str = obj["key"]

            if delimiter:
                suffix = key[len(prefix) :]
                delim_pos = suffix.find(delimiter)
                if delim_pos >= 0:
                    cp = prefix + suffix[: delim_pos + len(delimiter)]
                    if cp not in seen_prefixes:
                        seen_prefixes.add(cp)
                        common_prefixes.append(cp)
                    continue

            contents.append(obj)

        total_returned = len(contents) + len(common_prefixes)
        is_truncated = total_returned >= max_keys and len(objects) > total_returned

        next_continuation_token: str | None = None
        next_marker: str | None = None
        if is_truncated:
            if contents:
                last_key = contents[-1]["key"]
            elif common_prefixes:
                last_key = common_prefixes[-1]
            else:
                last_key = ""
            if last_key:
                next_continuation_token = last_key
                next_marker = last_key

        return {
            "contents": contents,
            "common_prefixes": sorted(common_prefixes),
            "is_truncated": is_truncated,
            "next_continuation_token": next_continuation_token,
            "next_marker": next_marker,
            "key_count": total_returned,
        }

    async def create_multipart_upload(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        content_type: str = "application/octet-stream",
        content_encoding: str | None = None,
        content_language: str | None = None,
        content_disposition: str | None = None,
        cache_control: str | None = None,
        expires: str | None = None,
        storage_class: str = "STANDARD",
        acl: str = "{}",
        user_metadata: str = "{}",
        owner_id: str = "",
        owner_display: str = "",
    ) -> None:
        self._uploads[upload_id] = {
            "upload_id": upload_id,
            "bucket": bucket,
            "key": key,
            "content_type": content_type,
            "content_encoding": content_encoding,
            "content_language": content_language,
            "content_disposition": content_disposition,
            "cache_control": cache_control,
            "expires": expires,
            "storage_class": storage_class,
            "acl": acl,
            "user_metadata": user_metadata,
            "owner_id": owner_id,
            "owner_display": owner_display,
            "initiated_at": _now_iso(),
        }
        if upload_id not in self._parts:
            self._parts[upload_id] = {}

    async def get_multipart_upload(
        self, bucket: str, key: str, upload_id: str
    ) -> dict[str, Any] | None:
        upload = self._uploads.get(upload_id)
        if upload and upload["bucket"] == bucket and upload["key"] == key:
            return upload
        return None

    async def complete_multipart_upload(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        size: int,
        etag: str,
        content_type: str = "application/octet-stream",
        content_encoding: str | None = None,
        content_language: str | None = None,
        content_disposition: str | None = None,
        cache_control: str | None = None,
        expires: str | None = None,
        storage_class: str = "STANDARD",
        acl: str = "{}",
        user_metadata: str = "{}",
    ) -> None:
        await self.put_object(
            bucket=bucket,
            key=key,
            size=size,
            etag=etag,
            content_type=content_type,
            content_encoding=content_encoding,
            content_language=content_language,
            content_disposition=content_disposition,
            cache_control=cache_control,
            expires=expires,
            storage_class=storage_class,
            acl=acl,
            user_metadata=user_metadata,
        )
        self._parts.pop(upload_id, None)
        self._uploads.pop(upload_id, None)

    async def abort_multipart_upload(self, bucket: str, key: str, upload_id: str) -> None:
        self._parts.pop(upload_id, None)
        self._uploads.pop(upload_id, None)

    async def put_part(
        self,
        upload_id: str,
        part_number: int,
        size: int,
        etag: str,
    ) -> None:
        if upload_id not in self._parts:
            self._parts[upload_id] = {}
        self._parts[upload_id][part_number] = {
            "upload_id": upload_id,
            "part_number": part_number,
            "size": size,
            "etag": etag,
            "last_modified": _now_iso(),
        }

    async def get_parts_for_completion(self, upload_id: str) -> list[dict[str, Any]]:
        parts = self._parts.get(upload_id, {})
        return sorted(parts.values(), key=lambda p: p["part_number"])

    async def list_parts(
        self,
        upload_id: str,
        part_number_marker: int = 0,
        max_parts: int = 1000,
    ) -> dict[str, Any]:
        all_parts = self._parts.get(upload_id, {})
        parts = [p for p in all_parts.values() if p["part_number"] > part_number_marker]
        parts = sorted(parts, key=lambda p: p["part_number"])[: max_parts + 1]

        result_parts = parts[:max_parts]
        is_truncated = len(parts) > max_parts
        next_marker = result_parts[-1]["part_number"] if is_truncated and result_parts else None

        return {
            "parts": result_parts,
            "is_truncated": is_truncated,
            "next_part_number_marker": next_marker,
        }

    async def list_multipart_uploads(
        self,
        bucket: str,
        prefix: str = "",
        delimiter: str = "",
        max_uploads: int = 1000,
        key_marker: str = "",
        upload_id_marker: str = "",
    ) -> dict[str, Any]:
        uploads = [u for u in self._uploads.values() if u["bucket"] == bucket]

        if prefix:
            uploads = [u for u in uploads if u["key"].startswith(prefix)]

        if key_marker:
            if upload_id_marker:
                uploads = [
                    u
                    for u in uploads
                    if u["key"] > key_marker
                    or (u["key"] == key_marker and u["upload_id"] > upload_id_marker)
                ]
            else:
                uploads = [u for u in uploads if u["key"] > key_marker]

        uploads = sorted(uploads, key=lambda u: (u["key"], u["initiated_at"]))

        result_uploads: list[dict[str, Any]] = []
        common_prefixes: list[str] = []
        seen_prefixes: set[str] = set()

        for upload in uploads:
            if len(result_uploads) + len(common_prefixes) >= max_uploads:
                break

            key = upload["key"]

            if delimiter:
                suffix = key[len(prefix) :]
                delim_pos = suffix.find(delimiter)
                if delim_pos >= 0:
                    cp = prefix + suffix[: delim_pos + len(delimiter)]
                    if cp not in seen_prefixes:
                        seen_prefixes.add(cp)
                        common_prefixes.append(cp)
                    continue

            result_uploads.append(upload)

        total = len(result_uploads) + len(common_prefixes)
        is_truncated = len(uploads) > total and total >= max_uploads

        next_key_marker: str | None = None
        next_upload_id_marker: str | None = None
        if is_truncated and result_uploads:
            last = result_uploads[-1]
            next_key_marker = last["key"]
            next_upload_id_marker = last["upload_id"]

        return {
            "uploads": result_uploads,
            "common_prefixes": sorted(common_prefixes),
            "is_truncated": is_truncated,
            "next_key_marker": next_key_marker,
            "next_upload_id_marker": next_upload_id_marker,
        }

    async def get_credential(self, access_key_id: str) -> dict[str, Any] | None:
        cred = self._credentials.get(access_key_id)
        if cred and cred.get("active", 1) == 1:
            return cred
        return None

    async def put_credential(
        self,
        access_key_id: str,
        secret_key: str,
        owner_id: str = "",
        display_name: str = "",
    ) -> None:
        self._credentials[access_key_id] = {
            "access_key_id": access_key_id,
            "secret_key": secret_key,
            "owner_id": owner_id,
            "display_name": display_name,
            "active": 1,
            "created_at": _now_iso(),
        }

    async def count_objects(self, bucket: str) -> int:
        return sum(1 for (b, _) in self._objects if b == bucket)

    async def reap_expired_uploads(self, ttl_seconds: int = 604800) -> list[dict]:
        from datetime import timedelta

        cutoff = datetime.now(timezone.utc) - timedelta(seconds=ttl_seconds)
        cutoff_str = cutoff.strftime("%Y-%m-%dT%H:%M:%S.000Z")

        reaped = []
        for upload_id, upload in list(self._uploads.items()):
            if upload["initiated_at"] < cutoff_str:
                reaped.append(
                    {
                        "upload_id": upload_id,
                        "bucket": upload["bucket"],
                        "key": upload["key"],
                    }
                )
                self._parts.pop(upload_id, None)
                del self._uploads[upload_id]

        return reaped

File: golang/internal/metadata/memory.go
package metadata

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"sync"
	"time"
)

type MemoryStore struct {
	mu          sync.RWMutex
	buckets     map[string]*BucketRecord
	objects     map[string]map[string]*ObjectRecord
	uploads     map[string]*MultipartUploadRecord
	parts       map[string]map[int]*PartRecord
	credentials map[string]*CredentialRecord
}

func NewMemoryStore() *MemoryStore {
	return &MemoryStore{
		buckets:     make(map[string]*BucketRecord),
		objects:     make(map[string]map[string]*ObjectRecord),
		uploads:     make(map[string]*MultipartUploadRecord),
		parts:       make(map[string]map[int]*PartRecord),
		credentials: make(map[string]*CredentialRecord),
	}
}

func (s *MemoryStore) Ping(ctx context.Context) error {
	return nil
}

func (s *MemoryStore) Close() error {
	return nil
}

func (s *MemoryStore) CreateBucket(ctx context.Context, bucket *BucketRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.buckets[bucket.Name]; exists {
		return fmt.Errorf("bucket already exists: %s", bucket.Name)
	}

	bucketCopy := *bucket
	if bucketCopy.ACL == nil {
		bucketCopy.ACL = json.RawMessage("{}")
	}
	s.buckets[bucket.Name] = &bucketCopy
	return nil
}

func (s *MemoryStore) GetBucket(ctx context.Context, name string) (*BucketRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	bucket, exists := s.buckets[name]
	if !exists {
		return nil, nil
	}
	bucketCopy := *bucket
	return &bucketCopy, nil
}

func (s *MemoryStore) DeleteBucket(ctx context.Context, name string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.buckets[name]; !exists {
		return fmt.Errorf("bucket not found: %s", name)
	}

	if objects, exists := s.objects[name]; exists && len(objects) > 0 {
		return fmt.Errorf("bucket not empty: %s", name)
	}

	for _, upload := range s.uploads {
		if upload.Bucket == name {
			return fmt.Errorf("bucket not empty: %s", name)
		}
	}

	delete(s.buckets, name)
	return nil
}

func (s *MemoryStore) ListBuckets(ctx context.Context, owner string) ([]BucketRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var buckets []BucketRecord
	for _, bucket := range s.buckets {
		if bucket.OwnerID == owner {
			bucketCopy := *bucket
			buckets = append(buckets, bucketCopy)
		}
	}

	sort.Slice(buckets, func(i, j int) bool {
		return buckets[i].Name < buckets[j].Name
	})

	return buckets, nil
}

func (s *MemoryStore) BucketExists(ctx context.Context, name string) (bool, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	_, exists := s.buckets[name]
	return exists, nil
}

func (s *MemoryStore) UpdateBucketAcl(ctx context.Context, name string, acl json.RawMessage) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	bucket, exists := s.buckets[name]
	if !exists {
		return fmt.Errorf("bucket not found: %s", name)
	}

	bucket.ACL = acl
	return nil
}

func (s *MemoryStore) PutObject(ctx context.Context, obj *ObjectRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.buckets[obj.Bucket]; !exists {
		return fmt.Errorf("bucket not found: %s", obj.Bucket)
	}

	if s.objects[obj.Bucket] == nil {
		s.objects[obj.Bucket] = make(map[string]*ObjectRecord)
	}

	objCopy := *obj
	if objCopy.ContentType == "" {
		objCopy.ContentType = "application/octet-stream"
	}
	if objCopy.StorageClass == "" {
		objCopy.StorageClass = "STANDARD"
	}
	if objCopy.ACL == nil {
		objCopy.ACL = json.RawMessage("{}")
	}
	if objCopy.UserMetadata == nil {
		objCopy.UserMetadata = make(map[string]string)
	}

	s.objects[obj.Bucket][obj.Key] = &objCopy
	return nil
}

func (s *MemoryStore) GetObject(ctx context.Context, bucket, key string) (*ObjectRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if bucketObjects, exists := s.objects[bucket]; exists {
		if obj, exists := bucketObjects[key]; exists {
			objCopy := *obj
			return &objCopy, nil
		}
	}
	return nil, nil
}

func (s *MemoryStore) DeleteObject(ctx context.Context, bucket, key string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if bucketObjects, exists := s.objects[bucket]; exists {
		delete(bucketObjects, key)
	}
	return nil
}

func (s *MemoryStore) ObjectExists(ctx context.Context, bucket, key string) (bool, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if bucketObjects, exists := s.objects[bucket]; exists {
		_, exists = bucketObjects[key]
		return exists, nil
	}
	return false, nil
}

func (s *MemoryStore) DeleteObjectsMeta(ctx context.Context, bucket string, keys []string) ([]string, []error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	var deleted []string
	var errs []error

	bucketObjects, exists := s.objects[bucket]
	if !exists {
		return keys, nil
	}

	for _, key := range keys {
		delete(bucketObjects, key)
		deleted = append(deleted, key)
	}

	return deleted, errs
}

func (s *MemoryStore) UpdateObjectAcl(ctx context.Context, bucket, key string, acl json.RawMessage) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if bucketObjects, exists := s.objects[bucket]; exists {
		if obj, exists := bucketObjects[key]; exists {
			obj.ACL = acl
			return nil
		}
	}
	return fmt.Errorf("object not found: %s/%s", bucket, key)
}

func (s *MemoryStore) ListObjects(ctx context.Context, bucket string, opts ListObjectsOptions) (*ListObjectsResult, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	maxKeys := opts.MaxKeys
	if maxKeys <= 0 {
		maxKeys = 1000
	}

	startAfter := opts.StartAfter
	if opts.ContinuationToken != "" {
		startAfter = opts.ContinuationToken
	}
	if opts.Marker != "" && startAfter == "" {
		startAfter = opts.Marker
	}

	var allObjects []ObjectRecord
	bucketObjects, exists := s.objects[bucket]
	if !exists {
		return &ListObjectsResult{}, nil
	}

	for _, obj := range bucketObjects {
		if opts.Prefix != "" && !strings.HasPrefix(obj.Key, opts.Prefix) {
			continue
		}
		if startAfter != "" && obj.Key <= startAfter {
			continue
		}
		objCopy := *obj
		allObjects = append(allObjects, objCopy)
	}

	sort.Slice(allObjects, func(i, j int) bool {
		return allObjects[i].Key < allObjects[j].Key
	})

	if opts.Delimiter == "" {
		isTruncated := len(allObjects) > maxKeys
		if isTruncated {
			allObjects = allObjects[:maxKeys]
		}
		result := &ListObjectsResult{
			Objects:     allObjects,
			IsTruncated: isTruncated,
		}
		if isTruncated && len(allObjects) > 0 {
			lastKey := allObjects[len(allObjects)-1].Key
			result.NextMarker = lastKey
			result.NextContinuationToken = lastKey
		}
		return result, nil
	}

	var objects []ObjectRecord
	prefixSet := make(map[string]bool)

	for _, obj := range allObjects {
		keyAfterPrefix := obj.Key
		if opts.Prefix != "" {
			keyAfterPrefix = obj.Key[len(opts.Prefix):]
		}

		delimIdx := strings.Index(keyAfterPrefix, opts.Delimiter)
		if delimIdx >= 0 {
			commonPrefix := opts.Prefix + keyAfterPrefix[:delimIdx+len(opts.Delimiter)]
			prefixSet[commonPrefix] = true
		} else {
			objects = append(objects, obj)
		}
	}

	var commonPrefixes []string
	for p := range prefixSet {
		commonPrefixes = append(commonPrefixes, p)
	}
	sort.Strings(commonPrefixes)

	totalEntries := len(objects) + len(commonPrefixes)
	isTruncated := totalEntries > maxKeys

	if isTruncated {
		type entry struct {
			key      string
			isPrefix bool
		}
		var entries []entry
		for _, obj := range objects {
			entries = append(entries, entry{key: obj.Key, isPrefix: false})
		}
		for _, p := range commonPrefixes {
			entries = append(entries, entry{key: p, isPrefix: true})
		}
		sort.Slice(entries, func(i, j int) bool {
			return entries[i].key < entries[j].key
		})

		if len(entries) > maxKeys {
			entries = entries[:maxKeys]
		}

		objects = nil
		prefixSet = make(map[string]bool)
		for _, e := range entries {
			if e.isPrefix {
				prefixSet[e.key] = true
			} else {
				for _, obj := range allObjects {
					if obj.Key == e.key {
						objects = append(objects, obj)
						break
					}
				}
			}
		}
		commonPrefixes = nil
		for p := range prefixSet {
			commonPrefixes = append(commonPrefixes, p)
		}
		sort.Strings(commonPrefixes)
	}

	result := &ListObjectsResult{
		Objects:        objects,
		CommonPrefixes: commonPrefixes,
		IsTruncated:    isTruncated,
	}
	if isTruncated {
		var lastKey string
		if len(objects) > 0 {
			lastKey = objects[len(objects)-1].Key
		}
		if len(commonPrefixes) > 0 {
			lastPrefix := commonPrefixes[len(commonPrefixes)-1]
			if lastPrefix > lastKey {
				lastKey = lastPrefix
			}
		}
		result.NextMarker = lastKey
		result.NextContinuationToken = lastKey
	}

	return result, nil
}

func (s *MemoryStore) CreateMultipartUpload(ctx context.Context, upload *MultipartUploadRecord) (string, error) {
	uploadID := upload.UploadID
	if uploadID == "" {
		var err error
		uploadID, err = generateUploadID()
		if err != nil {
			return "", err
		}
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.buckets[upload.Bucket]; !exists {
		return "", fmt.Errorf("bucket not found: %s", upload.Bucket)
	}

	uploadCopy := *upload
	uploadCopy.UploadID = uploadID
	if uploadCopy.ContentType == "" {
		uploadCopy.ContentType = "application/octet-stream"
	}
	if uploadCopy.StorageClass == "" {
		uploadCopy.StorageClass = "STANDARD"
	}
	if uploadCopy.ACL == nil {
		uploadCopy.ACL = json.RawMessage("{}")
	}
	if uploadCopy.UserMetadata == nil {
		uploadCopy.UserMetadata = make(map[string]string)
	}

	s.uploads[uploadID] = &uploadCopy
	return uploadID, nil
}

func (s *MemoryStore) GetMultipartUpload(ctx context.Context, bucket, key, uploadID string) (*MultipartUploadRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	upload, exists := s.uploads[uploadID]
	if !exists || upload.Bucket != bucket || upload.Key != key {
		return nil, nil
	}

	uploadCopy := *upload
	return &uploadCopy, nil
}

func (s *MemoryStore) PutPart(ctx context.Context, part *PartRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.uploads[part.UploadID]; !exists {
		return fmt.Errorf("upload not found: %s", part.UploadID)
	}

	if s.parts[part.UploadID] == nil {
		s.parts[part.UploadID] = make(map[int]*PartRecord)
	}

	partCopy := *part
	s.parts[part.UploadID][part.PartNumber] = &partCopy
	return nil
}

func (s *MemoryStore) ListParts(ctx context.Context, uploadID string, opts ListPartsOptions) (*ListPartsResult, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	maxParts := opts.MaxParts
	if maxParts <= 0 {
		maxParts = 1000
	}

	uploadParts, exists := s.parts[uploadID]
	if !exists {
		return &ListPartsResult{}, nil
	}

	var parts []PartRecord
	for pn, part := range uploadParts {
		if pn <= opts.PartNumberMarker {
			continue
		}
		partCopy := *part
		parts = append(parts, partCopy)
	}

	sort.Slice(parts, func(i, j int) bool {
		return parts[i].PartNumber < parts[j].PartNumber
	})

	isTruncated := len(parts) > maxParts
	if isTruncated {
		parts = parts[:maxParts]
	}

	result := &ListPartsResult{
		Parts:       parts,
		IsTruncated: isTruncated,
	}
	if isTruncated && len(parts) > 0 {
		result.NextPartNumberMarker = parts[len(parts)-1].PartNumber
	}

	return result, nil
}

func (s *MemoryStore) GetPartsForCompletion(ctx context.Context, uploadID string, partNumbers []int) ([]PartRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	uploadParts, exists := s.parts[uploadID]
	if !exists {
		return nil, nil
	}

	var parts []PartRecord
	for _, pn := range partNumbers {
		if part, exists := uploadParts[pn]; exists {
			partCopy := *part
			parts = append(parts, partCopy)
		}
	}

	sort.Slice(parts, func(i, j int) bool {
		return parts[i].PartNumber < parts[j].PartNumber
	})

	return parts, nil
}

func (s *MemoryStore) CompleteMultipartUpload(ctx context.Context, bucket, key, uploadID string, obj *ObjectRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.uploads[uploadID]; !exists {
		return fmt.Errorf("upload not found: %s", uploadID)
	}

	if s.objects[obj.Bucket] == nil {
		s.objects[obj.Bucket] = make(map[string]*ObjectRecord)
	}

	objCopy := *obj
	if objCopy.ContentType == "" {
		objCopy.ContentType = "application/octet-stream"
	}
	if objCopy.StorageClass == "" {
		objCopy.StorageClass = "STANDARD"
	}
	if objCopy.ACL == nil {
		objCopy.ACL = json.RawMessage("{}")
	}
	if objCopy.UserMetadata == nil {
		objCopy.UserMetadata = make(map[string]string)
	}

	s.objects[obj.Bucket][obj.Key] = &objCopy

	delete(s.parts, uploadID)
	delete(s.uploads, uploadID)

	return nil
}

func (s *MemoryStore) AbortMultipartUpload(ctx context.Context, bucket, key, uploadID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	upload, exists := s.uploads[uploadID]
	if !exists || upload.Bucket != bucket || upload.Key != key {
		return fmt.Errorf("upload not found: %s", uploadID)
	}

	delete(s.parts, uploadID)
	delete(s.uploads, uploadID)

	return nil
}

func (s *MemoryStore) ListMultipartUploads(ctx context.Context, bucket string, opts ListUploadsOptions) (*ListUploadsResult, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	maxUploads := opts.MaxUploads
	if maxUploads <= 0 {
		maxUploads = 1000
	}

	var allUploads []MultipartUploadRecord
	for _, upload := range s.uploads {
		if upload.Bucket != bucket {
			continue
		}
		if opts.Prefix != "" && !strings.HasPrefix(upload.Key, opts.Prefix) {
			continue
		}
		if opts.KeyMarker != "" {
			if upload.Key < opts.KeyMarker {
				continue
			}
			if upload.Key == opts.KeyMarker && opts.UploadIDMarker != "" && upload.UploadID <= opts.UploadIDMarker {
				continue
			}
		}
		uploadCopy := *upload
		allUploads = append(allUploads, uploadCopy)
	}

	sort.Slice(allUploads, func(i, j int) bool {
		if allUploads[i].Key != allUploads[j].Key {
			return allUploads[i].Key < allUploads[j].Key
		}
		return allUploads[i].InitiatedAt.Before(allUploads[j].InitiatedAt)
	})

	isTruncated := len(allUploads) > maxUploads
	if isTruncated {
		allUploads = allUploads[:maxUploads]
	}

	result := &ListUploadsResult{
		Uploads:     allUploads,
		IsTruncated: isTruncated,
	}
	if isTruncated && len(allUploads) > 0 {
		last := allUploads[len(allUploads)-1]
		result.NextKeyMarker = last.Key
		result.NextUploadIDMarker = last.UploadID
	}

	return result, nil
}

func (s *MemoryStore) GetCredential(ctx context.Context, accessKeyID string) (*CredentialRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	cred, exists := s.credentials[accessKeyID]
	if !exists {
		return nil, nil
	}

	credCopy := *cred
	return &credCopy, nil
}

func (s *MemoryStore) PutCredential(ctx context.Context, cred *CredentialRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	credCopy := *cred
	s.credentials[cred.AccessKeyID] = &credCopy
	return nil
}

func (s *MemoryStore) ReapExpiredUploads(ttlSeconds int) ([]ExpiredUpload, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	cutoff := time.Now().Add(-time.Duration(ttlSeconds) * time.Second)
	var expired []ExpiredUpload

	for uploadID, upload := range s.uploads {
		if upload.InitiatedAt.Before(cutoff) {
			expired = append(expired, ExpiredUpload{
				UploadID:   uploadID,
				BucketName: upload.Bucket,
				ObjectKey:  upload.Key,
			})
			delete(s.parts, uploadID)
			delete(s.uploads, uploadID)
		}
	}

	return expired, nil
}

func generateMemoryUploadID() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("generating upload ID: %w", err)
	}
	return hex.EncodeToString(b), nil
}
